tools/pubsys/src/repo/validate_repo/mod.rs

```diff
@@ -0,0 +1,228 @@
+    percent_target_files: u8,
+}
+
+fn attempt_read_target<T>(
+    repo: &Repository<'_, T>,
+    target: String,
+) -> Result<Option<impl Read>, repo_error::Error>
+where
+    T: tough::Transport,
+{
+    let max_attempts = 5;
+    for attempt in 1..=max_attempts {
+        match repo.read_target(&target) {
+            Err(err) => match err {
+                // Retry if it's a transport failure, might be due to intermittent connection issues
+                tough::error::Error::Transport { .. } => {
+                    error!(
+                        "Repo target fetch failure on attempt {} out of {}",
+                        attempt, max_attempts
+                    );
+                    if attempt >= max_attempts {
+                        return Err(repo_error::Error::ReadTarget {
+                            target,
+                            source: err,
+                        });
+                    }
+                }
+                // For other errors we should just return immediately and not retry
+                err => {
+                    return Err(repo_error::Error::ReadTarget {
+                        target,
+                        source: err,
+                    })
+                }
+            },
+            Ok(target_reader) => return Ok(target_reader),
+        };
+    }
+
+    Ok(None)
+}
+
+fn attempt_download_target(mut reader: impl Read, target: String) -> Result<(), Error> {
+    let max_attempts = 5;
+    for attempt in 1..=max_attempts {
+        info!("Downloading target: {}", target);
+        if let Err(err) = io::copy(&mut reader, &mut io::sink()) {
+            error!(
+                "Download error on attempt {} out of {}: {}",
+                attempt, max_attempts, err
+            );
+            if attempt >= max_attempts {
+                return Err(error::Error::TargetDownload {
+                    target,
+                    source: err,
+                });
+            }
+            continue;
+        }
+        // Successfully downloaded target
+        break;
+    }
+
+    Ok(())
+}
+
+/// Randomly samples specified percentage of listed targets in the TUF repo and tries to retrieve them.
+fn retrieve_percentage_of_targets<T>(repo: &Repository<'_, T>, percentage: u8) -> Result<(), Error>
+where
+    T: tough::Transport,
+{
+    let targets = &repo.targets().signed.targets;
+    let percentage = f32::from(percentage) / 100.0;
+    let num_to_retrieve = (targets.len() as f32 * percentage).ceil();
+    let mut rng = &mut rand::thread_rng();
+    let mut sampled_targets: Vec<String> = targets.keys().map(|key| key.to_string()).collect();
+    sampled_targets = sampled_targets
+        .choose_multiple(&mut rng, num_to_retrieve as usize)
+        .cloned()
+        .collect();
+
+    for target in sampled_targets {
+        let maybe_target_reader = attempt_read_target(repo, target.to_string())?;
+        let reader = maybe_target_reader.context(error::TargetMissing {
```

Comment:
Do we need the separate variables rather than calling `.context` directly on the result of `attempt_read_target`?  I think that's more standard, unless you're doing something special with the Err case.